### PR TITLE
mxnet: use openblas, fix build

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change
noticed it was failing when reviewing https://github.com/NixOS/nixpkgs/pull/87102

previous failure:
```
builder for '/nix/store/1zcd9mmm3dyyxcr588cszdclrsnjbga1-mxnet-1.6.0.drv' failed with exit code 1; last 10 log lines:
  -- Could not find OpenBLAS lib. Turning OpenBLAS_FOUND off
  CMake Error at cmake/Modules/FindOpenBLAS.cmake:82 (MESSAGE):
    Could not find OpenBLAS
  Call Stack (most recent call first):
    cmake/ChooseBlas.cmake:42 (find_package)
    CMakeLists.txt:310 (include)


  -- Configuring incomplete, errors occurred!
  See also "/build/apache-mxnet-src-1.6.0-incubating/build/CMakeFiles/CMakeOutput.log".
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/math/mxnet/default.nix
Comment (bhipple) [inline pkgs/applications/science/math/mxnet/default.nix:2]:
```suggestion
, opencv3, gtest, blas, perl
```

File: pkgs/applications/science/math/mxnet/default.nix
Comment (bhipple) [inline pkgs/applications/science/math/mxnet/default.nix:20]:
```suggestion
  buildInputs = [ opencv3 gtest blas blas.provider ]
```

**Review by bhipple — CHANGES_REQUESTED** [review]
We should keep using the virtual package, so that if someone overlays `mkl` as the `blas` implementation `mxnet` will pick it up correctly.  

Sometimes, we need the specific provider package in buildInputs as well, to get things like cmake hooks or path searches working.  The way to do that is to also add `blas.provider` to the `buildInputs`, which will cleanly and consistently handle the choice.

CC @matthewbauer I'm pretty sure the above is correct; I had to do something similar for `pytorch`.  Perhaps we should add to the manual?

File: pkgs/applications/science/math/mxnet/default.nix
Comment (jonringer) [reply pkgs/applications/science/math/mxnet/default.nix:20]: ah, wasn't aware of the provider usage

Comment (jonringer) [reply pkgs/applications/science/math/mxnet/default.nix:20]: is there any need for the actual `blas` package/attr set? Seems to build fine with only the provider

Comment (jonringer) [reply pkgs/applications/science/math/mxnet/default.nix]: oh, linking maybe

Comment (bhipple) [reply pkgs/applications/science/math/mxnet/default.nix]: Build just failed for me; looks like we need to do the same with `lapack` to make it visible to the linker.

Comment (bhipple) [reply pkgs/applications/science/math/mxnet/default.nix]: > is there any need for the actual blas package/attr set? Seems to build fine with only the provider

Not actually sure ... in principle it might be sufficient to just pass in `blas.provider`, but it seems reasonable to expose the meta-package as well, as that has the actual public API shared objects defined.

**Review by bhipple — APPROVED** [review]
`mxnet` built on my machine with this update now; kicking off a `nixpkgs-review` now.

**Comment by jonringer** [comment]
I'm not a big fan of having to do both `blas blas.provider`, is there a way that we can `symLinkJoin` them so that people only need to specify `blas`

**Comment by jonringer** [comment]
AFAIK, the `provider` pass thru is pretty special to blas libraries

**Comment by bhipple** [comment]
Result of `nixpkgs-review pr 87118` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
<br>- python27Packages.myfitnesspal
</details>
<details>
  <summary>5 packages built:</summary>
<br>- mxnet
<br>- python27Packages.mxnet
<br>- python37Packages.mxnet
<br>- python37Packages.optuna
<br>- python38Packages.mxnet
</details>

**Comment by jonringer** [comment]
I would approve your push, but I can't approve my own PR. Lol

**Comment by jonringer** [comment]
also, wouldn't it be better to get the `liblapack.so` from the `blas.provider`?

```
[12:24:46] jon@jon-desktop /home/jon/projects/nixpkgs (fix-mxnet)
$ nix-build -A blas.provider
/nix/store/hwdbqpqknffqnc30fjxzvrpmp2d3lcg2-openblas-0.3.9
[12:24:58] jon@jon-desktop /home/jon/projects/nixpkgs (fix-mxnet)
$ ls ./result/lib/lib*
./result/lib/libblas.so.3     ./result/lib/liblapack.so.3          ./result/lib/libopenblas.so.0
./result/lib/libcblas.so.3    ./result/lib/libopenblasp-r0.3.9.so
./result/lib/liblapacke.so.3  ./result/lib/libopenblas.so
```

**Comment by jonringer** [comment]
the main thing I see, is that the `blas.provider` only provides the libraries with the `.3` suffix, maybe we should add symlinks to their unversioned variants

**Comment by bhipple** [comment]
Agreed, the current setup is a little clunky, though I'm not sure `symlinkJoin` would cover it: IIRC we need the actual package to be in `buildInputs`, so that setupHooks for doing things like adding to `PKG_CONFIG_PATH` and the CMake search path work.

@matthewbauer anything you think we should change here?

**Comment by jonringer** [comment]
actually, this is specific to `openblas`
```
    ln -s $out/lib/libopenblas${stdenv.hostPlatform.extensions.sharedLibrary} $out/lib/libblas${stdenv.hostPlatform.extensions.sharedLibrary}${stdenv.lib.optionalString stdenv.hostPlatform.isLinux ".3"}
    ln -s $out/lib/libopenblas${stdenv.hostPlatform.extensions.sharedLibrary} $out/lib/libcblas${stdenv.hostPlatform.extensions.sharedLibrary}${stdenv.lib.optionalString stdenv.hostPlatform.isLinux ".3"}
    ln -s $out/lib/libopenblas${stdenv.hostPlatform.extensions.sharedLibrary} $out/lib/liblapack${stdenv.hostPlatform.extensions.sharedLibrary}${stdenv.lib.optionalString stdenv.hostPlatform.isLinux ".3"}
    ln -s $out/lib/libopenblas${stdenv.hostPlatform.extensions.sharedLibrary} $out/lib/liblapacke${stdenv.hostPlatform.extensions.sharedLibrary}${stdenv.lib.optionalString stdenv.hostPlatform.isLinux ".3"}
 ```

**Comment by jonringer** [comment]
by modifying the `openblas` package (in the PR that's now linked to this thread), i was able to build this while only supplying `blas.provider`

**Comment by jonringer** [comment]
currently blocked by  #87134

**Comment by matthewbauer** [comment]
> CC @matthewbauer I'm pretty sure the above is correct; I had to do something similar for `pytorch`. Perhaps we should add to the manual?

Yes that sounds right. Something in the manual would be helpful. We may need a "maintainers" section or something since this is more about how you make sure your package is correct vs. just how you use it.

> @matthewbauer anything you think we should change here?

Adding the unversioned library seems like a good idea.

**Review by matthewbauer — APPROVED** [review]

**Comment by jonringer** [comment]
should I just target staging and get this merged?

**Comment by bhipple** [comment]
Sounds good, let's just merge this on staging then so it can bake with all the other openblas / numpy changes. and we don't have to remember to circle back to it.

**Comment by jonringer** [comment]
@GrahamcOfBorg build mxnet

**Comment by doronbehar** [comment]
There's a CI error.

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

**Comment by jonringer** [comment]
pretty sure staging was broken at the time of PR creation

**Pull request merged**
